e2e/framework.go

```diff
@@ -60,10 +60,73 @@ func initTracer(name string) *sdktrace.TracerProvider {
 }
 
 type cluster struct {
-	t      *testing.T
-	nodes  map[string]*node
-	tracer *sdktrace.TracerProvider
-	hook   transportHook
+	t               *testing.T
+	lock            sync.Mutex
+	nodes           map[string]*node
+	tracer          *sdktrace.TracerProvider
+	hook            transportHook
+	sealedProposals []*pbft.SealedProposal
+	proposer        pbft.NodeID
+}
+
+// getCurrentIndex returns a last index up to which there are inserted proposals
+func (c *cluster) getCurrentIndex() uint64 {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	idx := 0
+	if len(c.sealedProposals) > 0 {
+		idx = len(c.sealedProposals) - 1
+	}
+
+	return uint64(idx)
+}
+
+// getSyncIndex returns a index up to which the node is synced with the network
+func (c *cluster) getSyncIndex(node string) uint64 {
+	c.nodes[node].lock.Lock()
+	defer c.nodes[node].lock.Unlock()
+	return c.nodes[node].syncIndex
+}
+
+// getNodeHeight returns a node height for a given index
+func (c *cluster) getNodeHeight(index uint64) uint64 {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	height := uint64(1) // initial height is always 1 since 0 is the genesis
+	if len(c.sealedProposals) != 0 {
+		height = c.sealedProposals[index].Number
+	}
+	return height
+}
+
+// insertFinalProposal inserts final proposal from the node to the cluster and returns index up to which the node is synced
+func (c *cluster) insertFinalProposal(p *pbft.SealedProposal) uint64 {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for _, v := range c.sealedProposals {
```

Comment:
This is the safe check to validate that two proposals are the same. If all the nodes block at this function, shouldn't the proposal be always the last one?

Reply:
Removed since it is no longer required.